### DISCUSSION TITLE
SSH Key management

### DISCUSSION THREAD

**Discussion by orangedeng** [cnrancher/autok3s#539] [Ideas]
The SSH key management will have a CLI and UI to management the SSH keys you want to use to connect to your to-install k3s node.

Following data struct would be the SSH key object's:
|field | desc|
|---|---|
|name|the name of the SSH Keys|
|ssh-cert|the certificate signed by SSH CA with the public keys in it|
|ssh-key|the SSH private key|
|ssh-key-public|the SSH public key|

## SSH key management

With the SSH key management, you can do following operations:

- **List** the stored ssh key pairs
- **Create** a SSH key pair with a specific name. The private key is necessary and the public key, the SSH certificate and the key passphrase are optional for input.
- **Generate** a SSH key pair with the RSA algorithm. The key passphrase and the bits(default to 2048) are optional for input. The public and private key will be the output. You can also specify the directory for saving the key files and the names of the files would be `id_rsa` and `id_rsa.pub`.
- **Delete** a stored SSH key pair.

Both CLI and web UI have above functions.

## Cluster provision

In k3s cluster create option, you can specify `--ssh-key-name` to use the stored SSH key pair instead of `--ssh-key-path` and `--ssh-cert-path`.

When using the stored key to provision a new cluster, the keys will be copied to the cluster's directory and the path of the key and cert will be saved to the cluster record.

When provisioning a new cluster in **web UI**, following changes should be made:

- Support to select a stored ssh key pair to provision the cluster the field would be `ssh-key-name`. The key will show whether it requires passphrase or not.
- When using a raw private to provision a new cluster, the web UI should support user to select the private key from files and read the file content to send to backend via API. The field to send the file content is `ssh-key`.

## TODOs

In the feature release, we are going to support to check or create SSH key pair with the stored one in the provider so that we don't need to provide `--key-pair` parameters.

**Comment by ly5156** [comment]
![ssh-keys](https://user-images.githubusercontent.com/3406205/209650464-d3df25d6-7958-4074-b2db-3164c8bb7b42.gif)


![ssh-keys2](https://user-images.githubusercontent.com/3406205/209651754-1386047d-944d-4d48-bf7e-c1c4e079118e.gif)
